server/server.py

```diff
@@ -309,13 +314,39 @@ def get_dense_path():
     return "No keys found"
 
 
+@socketio.on_error()
+def socket_io_error_handler(e):
+    print('SocketIO error: ' + str(e))
+
+
+def redis_change_handler(msg):
+    key = msg['data'].decode()
+    if redis_utils.is_robot_key(key):
+        print(key)
+        data = robots_to_json([key])
+        socketio.emit('herd-data', data, callback=lambda x: print(x))
+
+
+def exception_handler(ex, pubsub, thread):
+    print(ex)
+    thread.stop()
+    thread.join(timeout=1.0)
+    pubsub.close()
+
+
 if __name__ == "__main__":
     while True:
         try:
-            app.run(debug=True,
-                    use_reloader=True,
-                    host="0.0.0.0",
-                    port=int("80"))
-        except BaseException:
+            pubsub = redis_client.pubsub()
+            # see https://redis.io/topics/notifications
+            pubsub.psubscribe(**{'__keyevent@0__:set': redis_change_handler})
```

Comment:
With more than one vehicles in Redis, some active and others inactive, it's critical to replace the previous `/get_herd_data` polling with the new SocketIO based active notification. It's based on http://redis.io/topics/notifications: when a Redis key is set by `zmq_server.py`, a new event is triggered and a message is broadcasted to all connected web clients. Inactive vehicles won't trigger such message hence not being shown up on the web UI.